### ISSUE TITLE
Feature request - Allow tab selection to send an OSC message

### ISSUE THREAD

**Issue by ovenwerks** [jean-emmanuel/open-stage-control#69]
I don't know if this has been asked before... but it would be quite nice if selecting a tab could send an arbitrary OSC message. In my case I would like to send a page size message... /send/page_size 8 (for 8send channels) or /plugin/page_size 32 for a tab that has 32 plugin controls (some plugins have 100s of controls)

**Comment by jean-emmanuel** [comment]
For now tabs don't behave like the other widgets and cannot send osc, but this is something I'd like to implement someday.

**Comment by ovenwerks** [comment]
Assuming this comment really belonged in here... 
" Now another question: is it possible to switch to a tab via OSC messages?

yep: http://osc.ammd.net/user-guide/remote-control/#tabs-id-id-etc"

It seems that I could use a button to do this, provided I could customize it some(or use two linked buttons on top of each other). I want to stay away from controller specific commands in Ardour's code. With targets, being able to have a different OSC path for each target might be appealing: send one path to the DAW and another back to the controller /tab 2 (or as above use two buttons).

This would beg the question, can I hide the tabs and replace them with my own buttons?  ;) (I am guessing there is some CSS command that would either shrink them to 1 or 2 pixels high or hide them, but my js abilities are not there.)

more thoughts on having tab, or other controller specific commands sent from the DAW. I think these would end up being triggered by the controller not the DAW. So the DAW would in effect be getting used as an OSC filter to take something the controller sent and send something back to the controller. In my mind that doesn't make sense. The controller should be able to take care of itself. The one exception I could see, would be for the DAW to be able to send a *.js at startup... but again this would mean setting up another server in the DAW... a lot of work that may be better replaced by a straight web interface. Sorry for the long comment.

**Comment by jean-emmanuel** [comment]
Short answer would be no, not yet, not until I refactor tabs to act like widgets, its one of the oldest part in the code ;)

But, in the mean time, as a workaround, you can make a button send the /TABS command to app itself; as for hiding the navbar height, you can add `--nav-height:0px;` to the panel's css

**Comment by jean-emmanuel** [comment]
Tab system rewrite is almost finished !!

**Comment by jean-emmanuel** [comment]
Ok, I just need to write some docs and do some tests, the release should come soon. There are breaking changes (see [changelog](https://github.com/jean-emmanuel/open-stage-control/blob/master/CHANGELOG.md#0180)).

**Comment by ovenwerks** [comment]
as of git 5aefbd5125136b1c7103d6906b64cdd4a3726826 selecting any of the tabs sends the root OSC message rather than the tab's OSC message :)

**Comment by jean-emmanuel** [comment]
Panels send osc messages when their own tabs are selected. Since tabs and root are panels too, their osc settings appliy the same way (the main tabs belong the root panel, they trigger its osc messages). I'll try to make that as clear as possible in the docs ;)

**Comment by ovenwerks** [comment]
So I press tab one and I send the root panel message? So how do I send the tab's message (the one I actually need) I have 5 tabs and they all send /root... but when I click on tab 3 I want it to send /select/send_page_size i 6 and when I click on tab 5 I want it to send /select/plug_page_size i 5 and I see no way of doing that. In my opinion /root should only be sent on session load... I guess I am for shrinking the tab and using stacked, linked buttons to change tabs and send the messages I need.

So another feature request... to make this switchable so clicking a tab sends the tabs message and not the root message...

**Comment by jean-emmanuel** [comment]
Your five tabs send /root indeed, but with different arguments. In o-s-c, a widget has an address and sends or receives its state described by the value. The state of a tabs is handled by its parent panel/tab, you don't activate a tab, you tell its parent to select it (the value being the selected tab's index). The point of rewriting the tab system was to avoid the exceptions it caused in its early form, I'd prefer not to reintroduce one here. 

With custom module, it's easy to do anyway, by filtering the app's osc output:

```js
(function(){

    // Do whatever you want, initialize some variables, declare some functions, ...

    return {
        init: function(){
            // this will be executed once when the osc server starts
        },
        oscInFilter:function(data){
            // Filter incomming osc messages

            var {address, args, host, port} = data

    

            // return data if you want the message to be processed
            return {address, args, host, port}

        },
        oscOutFilter:function(data){
            // Filter outgoing osc messages

            var {address, args, host, port} = data

            if (address == '/root') {
                switch (args[0].value) {
                    case 3:
                        sendOsc({
                            address:'/select/send_page_size',
                            args: [
                                {type:'i', value:6}
                            ]
                        })
                    case 5:
                        sendOsc({
                            address:'/select/send_page_size',
                            args: [
                                {type:'i', value:5}
                            ]
                        })            
                }
            }

            // return data if you want the message to be and sent
            return {address, args, host, port}
        }
    }

})()
```

**Comment by ovenwerks** [comment]
Ok, that makes sense. I should really be doing what I want to do another way anyway, I was being lazy. /set_surface now has 6 parameters adding send_page and plugin_page. This ends up being better and more consistent. /root i tab_number may be useful down the road too. Thank you.

**Comment by jean-emmanuel** [comment]
Ok cool ! Thanks for testing by the way :)

**Comment by jean-emmanuel** [comment]
Included in v0.18.0
